### PR TITLE
chore: address deprecations caused by updating packages

### PR DESCRIPTION
**Pull request by NiceAesth**
# Describe your changes

RFC. This is the minimal required to get it into a working state. See: https://fastapi.tiangolo.com/advanced/events/#lifespan

Breaking changes spotted:
- https://github.com/tiangolo/fastapi/releases/tag/0.106.0
- https://github.com/tiangolo/fastapi/releases/tag/0.93.0
- https://www.starlette.io/release-notes/#0260
- https://www.starlette.io/release-notes/#0340
- https://github.com/redis/redis-py/releases/tag/v5.0.1
- akatsuki-pp-py version bump. Did we historically handle pp system versioning in some way? @cmyui 

As a side effect, we now should technically support Python 3.12

Other things of note:
~~We may support Pytest 8 soon once pytest-asyncio v23.5 is out of alpha. https://github.com/pytest-dev/pytest-asyncio/issues/737#issuecomment-1930632667~~ done in #571 

## Related Issues / Projects

Depends on #568 

## Checklist
- [x] I've manually tested my code

### CODE REVIEW

File: app/api/init_api.py [hunk 1]
```diff
@@ -64,6 +67,57 @@ def openapi(self) -> dict[str, Any]:
         return self.openapi_schema
 
 
+@asynccontextmanager
+async def lifespan(asgi_app: BanchoAPI) -> AsyncIterator[Never]:
+    app.state.loop = asyncio.get_running_loop()
```
Comment (NiceAesth) [inline app/api/init_api.py:72]: potentially could get rid of state in the future and use the lifespan instead. worth discussing at the very least. left alone for now

File: tools/recalc.py [hunk 2]
```diff
@@ -263,7 +263,7 @@ async def main(argv: Sequence[str] | None = None) -> int:
 
     await app.state.services.http_client.aclose()
     await db.disconnect()
-    await redis.close()
+    await redis.aclose()
```
Comment (NiceAesth) [inline tools/recalc.py:266]: `close()` is deprecated as of 5.0.1

File: app/api/init_api.py [hunk 3]
```diff
@@ -64,6 +67,57 @@ def openapi(self) -> dict[str, Any]:
...
+
+    await app.bg_loops.initialize_housekeeping_tasks()
+
+    log("Startup process complete.", Ansi.LGREEN)
+    log(
+        f"Listening @ {app.settings.APP_HOST}:{app.settings.APP_PORT}",
+        Ansi.LMAGENTA,
+    )
+
+    yield  # type: ignore
```
[the remark this answers is not in the document]
Comment (cmyui) [reply app/api/init_api.py:104]: just want to make sure, is this a 1:1 copy other than the yield?

File: tests/integration/domains/osu_test.py [hunk 4]
```diff
@@ -239,5 +239,5 @@ async def test_score_submission(
...
+        == b"beatmapId:315|beatmapSetId:141|beatmapPlaycount:1|beatmapPasscount:1|approvedDate:2014-05-18 15:41:48|\n|chartId:beatmap|chartUrl:https://osu.cmyui.xyz/s/141|chartName:Beatmap Ranking|rankBefore:|rankAfter:1|rankedScoreBefore:|rankedScoreAfter:26810|totalScoreBefore:|totalScoreAfter:26810|maxComboBefore:|maxComboAfter:52|accuracyBefore:|accuracyAfter:81.94|ppBefore:|ppAfter:10.04103|onlineScoreId:1|\n|chartId:overall|chartUrl:https://cmyui.xyz/u/3|chartName:Overall Ranking|rankBefore:|rankAfter:1|rankedScoreBefore:|rankedScoreAfter:26810|totalScoreBefore:|totalScoreAfter:26810|maxComboBefore:|maxComboAfter:52|accuracyBefore:|accuracyAfter:81.94|ppBefore:|ppAfter:10|achievements-new:osu-skill-pass-4+Insanity Approaches+You're not twitching, you're just ready./all-intro-hidden+Blindsight+I can see just perfectly"
```
[the remark this answers is not in the document]
Comment (cmyui) [reply tests/integration/domains/osu_test.py:242]: we should prolly refactor this to use snapshot tests

[the remark this answers is not in the document]
Comment (cmyui) [reply tests/integration/domains/osu_test.py:242]: also the test suite runs on the docker db rn which is super weird - we should be spinning up a temp db & running migrations every time we run tests

**Review by cmyui — APPROVED** [review]

**Comment by cmyui** [comment]
+1 👍🏻

File: app/api/init_api.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (NiceAesth) [reply app/api/init_api.py:104]: Yep

File: app/api/init_api.py [hunk 5]
```diff
@@ -64,6 +67,57 @@ def openapi(self) -> dict[str, Any]:
         return self.openapi_schema
 
 
+@asynccontextmanager
+async def lifespan(asgi_app: BanchoAPI) -> AsyncIterator[Never]:
```
Comment (cmyui) [inline app/api/init_api.py:71]: should this be AsyncIterator[None]?

Comment (NiceAesth) [reply app/api/init_api.py:71]: i remember trying that and it failing mypy

**Pull request merged**